fix validators run during fits array writing

braingram opened the pull request:
Closes #279

This PR fixes the ndim, max_ndim and datatype validation during FITS array writing by yielding from the corresponding validators (these return generators that are not consumed on main).

This PR also fixes the `if` check for `dtype` to instead check for `datatype` (the keyword used in the schema) and uses the custom `_validate_datatype` defined in this package (rather than `validate_dtype` from asdf which doesn't support `ndim` in structured datatypes or `allow_extra_columns`.

An update to one of the `test_multislit.py` unit tests was required where a float64 was being assigned to an array with a float32 dtype defined in the schema.

jwst regression tests all pass: https://github.com/spacetelescope/RegressionTests/actions/runs/13841630347/job/38773673047

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [ ] update or add relevant tests
- [ ] update relevant docstrings and / or `docs/` page
- [ ] Does this PR change any API used downstream? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] [run `jwst` regression tests](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) with this branch installed (`"git+https://github.com/<fork>/stdatamodels@<branch>"`)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.feature.rst``: new feature
- ``changes/<PR#>.bugfix.rst``: fixes an issue
- ``changes/<PR#>.doc.rst``: documentation change
- ``changes/<PR#>.removal.rst``: deprecation or removal of public API
- ``changes/<PR#>.misc.rst``: infrastructure or miscellaneous change
</details>

emolter reviewed (commented):
Looks good to me.  I'm a bit confused seeing that there were no regtest failures - didn't you say when we had a Slack huddle that there were going to be many failures in JWST?  What happened to that?

braingram commented:
> Looks good to me. I'm a bit confused seeing that there were no regtest failures - didn't you say when we had a Slack huddle that there were going to be many failures in JWST? What happened to that?

Magic! :)

It turns out that the assignment validation of array datatypes in stdatamodels differs from what's done in the standard asdf datatype validator. To see all the differences compare:
- https://github.com/spacetelescope/stdatamodels/blob/f504f64263473cb9b61f4bd8496179f0d2e082a0/src/stdatamodels/validate.py#L44
- https://github.com/asdf-format/asdf/blob/d70ac0fa103cb884a7d9fd4276e6a320b68aacb0/asdf/tags/core/ndarray.py#L517

This is already established usage since stdatamodels registers it's own `_validate_datatype` with the jsonschema validator:
https://github.com/spacetelescope/stdatamodels/blob/f504f64263473cb9b61f4bd8496179f0d2e082a0/src/stdatamodels/validate.py#L138
When I fixed the if here:
https://github.com/spacetelescope/stdatamodels/blob/f504f64263473cb9b61f4bd8496179f0d2e082a0/src/stdatamodels/fits_support.py#L291-L292
the standard asdf datatype validator was applied to the datatypes in the schemas (when datamodels were written). This resulted in a bunch of failures (since the stdatamodels datatype definitions don't match the asdf ones). Once I realized this and switched the above line to instead use the stdatamodels `_validate_datatype` the errors went away (except for the one in the unit test fixed in this PR).
